pytorch_lightning/loops/optimizer/optimizer_loop.py

```diff
@@ -361,18 +352,16 @@ def _training_step(
 
             _check_training_step_output(self.trainer.lightning_module, training_step_output)
 
-            result_collection, self._hiddens = _process_training_step_output(self.trainer, training_step_output)
-            if result_collection is None:
-                return None
+            closure_result = ClosureResult.from_training_step_output(training_step_output)
+
+            if self.trainer.terminate_on_nan:
+                check_finite_loss(closure_result.closure_loss)
 
-        # output validation already done, here loss can't be None
-        assert result_collection.minimize is not None
+            if self.trainer.move_metrics_to_cpu:
+                self.trainer._results.cpu()
 
-        # accumulate loss. if accumulate_grad_batches==1, no effect
-        closure_loss = result_collection.minimize / self.trainer.accumulate_grad_batches
-        # the loss will get scaled for amp. avoid any modifications to it
-        loss = closure_loss.detach().clone()
-        return AttributeDict(closure_loss=closure_loss, loss=loss, result_collection=result_collection)
+        closure_result.apply_accumulation(self.trainer.accumulate_grad_batches)
```

Comment:
Can't be done there if we want to avoid a trainer reference, however, the following is possible


```python
closure_result = ClosureResult.from_training_step_output(
    training_step_output,
    normalize=self.trainer.accumulate_grad_batches
)

@classmethod
def from_training_step_output(cls, training_step_output, normalize: int = 1):
    ...
    closure_loss /= normalize
    result = cls(closure_loss, hiddens, extra=extra)
```

Reply:
I would just pass `accumulate_grad_batches` without creating the closure. No need for a trainer reference.